File: examples/canfd_demo.cpp
// canfd_demo: end-to-end hardware demo for gsusb-canfd on macOS/Linux/Windows.
//
// It scans for an adapter, opens it, prints the negotiated bit timing, transmits
// a trigger frame on 0x501 and prints every response it sees (0x481 by default),
// skipping the adapter's own TX echo.
//
// Build:  cmake --build build --target canfd_demo
// Run:    ./build/canfd_demo --seconds 6
//         ./build/canfd_demo --monitor --seconds 5
//         ./build/canfd_demo --classic --id 123 --rx-id 456 --take-data 11223344

#include <algorithm>
#include <atomic>
#include <cctype>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <map>
#include <string>
#include <thread>
#include <vector>

#include "canfd/canfd.hpp"

namespace {

std::atomic<bool> g_stop{false};

void onSignal(int) { g_stop.store(true); }

std::string featureString(uint32_t feature) {
  std::string out;
  auto add = [&](const char* name) {
    if (!out.empty()) {
      out += "|";
    }
    out += name;
  };
  if (feature & canfd::feature::kListenOnly) add("listen_only");
  if (feature & canfd::feature::kLoopback) add("loopback");
  if (feature & canfd::feature::kOneShot) add("one_shot");
  if (feature & canfd::feature::kHwTimestamp) add("hw_timestamp");
  if (feature & canfd::feature::kFd) add("fd");
  if (feature & canfd::feature::kBtConstExt) add("bt_const_ext");
  if (out.empty()) {
    out = "none";
  }
  return out;
}

double samplePointOf(const canfd::BitTiming& t) {
  const double total = 1.0 + t.tseg1() + t.phase_seg2;
  return static_cast<double>(1 + t.tseg1()) / total;
}

double bitrateOf(const canfd::BitTiming& t, uint32_t fclk) {
  const double total = 1.0 + t.tseg1() + t.phase_seg2;
  return static_cast<double>(fclk) / (static_cast<double>(t.brp) * total);
}

void printTiming(const char* label, const canfd::BitTiming& t, uint32_t fclk) {
  std::printf("  %-8s brp=%u tseg1=%u tseg2=%u sjw=%u  ->  %.0f bit/s @ %.1f%%\n", label,
              t.brp, t.tseg1(), t.phase_seg2, t.sjw, bitrateOf(t, fclk), samplePointOf(t) * 100.0);
}

std::vector<uint8_t> parseData(const std::string& text) {
  std::string cleaned;
  for (char c : text) {
    if (std::isxdigit(static_cast<unsigned char>(c))) {
      cleaned += c;
    }
  }
  std::vector<uint8_t> out;
  for (std::size_t i = 0; i + 1 < cleaned.size(); i += 2) {
    out.push_back(static_cast<uint8_t>(std::strtoul(cleaned.substr(i, 2).c_str(), nullptr, 16)));
  }
  return out;
}

canfd::CanFrame makeFrame(uint32_t id, const std::vector<uint8_t>& data, bool fd) {
  canfd::CanFrame f;
  f.id = id;
  f.extended = id > 0x7FF;
  f.fd = fd;
  f.brs = fd;
  f.size = static_cast<uint8_t>(data.size());
  for (std::size_t i = 0; i < data.size() && i < canfd::kMaxPayload; ++i) {
    f.data[i] = data[i];
  }
  return f;
}

struct Args {
  double seconds = 6.0;
  bool monitor = false;
  bool classic = false;
  bool trigger = true;
  canfd::BusConfig config;
  uint32_t trigger_id = 0x501;
  std::vector<uint8_t> trigger_data = {0x00, 0x02, 0x50, 0x01};
  uint32_t rx_id = 0x481;
  bool filter_rx_id = true;
  bool print_all = false;
};

Args parseArgs(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    const std::string a = argv[i];
    auto next = [&]() -> std::string {
      if (i + 1 >= argc) {
        std::fprintf(stderr, "missing value for %s\n", a.c_str());
        std::exit(2);
      }
      return argv[++i];
    };
    if (a == "--seconds") {
      args.seconds = std::atof(next().c_str());
    } else if (a == "--monitor") {
      args.monitor = true;
      args.trigger = false;
    } else if (a == "--no-trigger") {
      args.trigger = false;
    } else if (a == "--classic") {
      args.classic = true;
      args.config.fd = false;
    } else if (a == "--bitrate") {
      args.config.bitrate = std::strtoul(next().c_str(), nullptr, 0);
    } else if (a == "--sample-point") {
      args.config.sample_point = std::atof(next().c_str());
    } else if (a == "--data-bitrate") {
      args.config.data_bitrate = std::strtoul(next().c_str(), nullptr, 0);
    } else if (a == "--data-sample-point") {
      args.config.data_sample_point = std::atof(next().c_str());
    } else if (a == "--id") {
      args.trigger_id = std::strtoul(next().c_str(), nullptr, 16);
    } else if (a == "--take-data") {
      args.trigger_data = parseData(next());
    } else if (a == "--rx-id") {
      args.rx_id = std::strtoul(next().c_str(), nullptr, 16);
    } else if (a == "--all") {
      args.filter_rx_id = false;
    } else if (a == "--all-frames") {
      args.print_all = true;
    } else {
      std::fprintf(stderr, "unknown option %s\n", a.c_str());
      std::exit(2);
    }
  }
  return args;
}

}  // namespace

int main(int argc, char** argv) {
  const Args args = parseArgs(argc, argv);
  std::signal(SIGINT, onSignal);

  std::printf("=== scan ===\n");
  for (const auto& a : canfd::scanAdapters(0, 0)) {
    std::printf("  %s | %s | %04X:%04X bus=%u addr=%u serial=%s\n", a.manufacturer.c_str(),
                a.product.c_str(), a.vendor_id, a.product_id, a.bus, a.address, a.serial.c_str());
  }

  canfd::CanFdBus bus;
  try {
    std::printf("=== open ===\n");
    bus.open();
    const auto& info = bus.deviceInfo();
    std::printf("  endpoints in=0x%02X out=0x%02X\n", bus.inputEndpoint(), bus.outputEndpoint());
    std::printf("  firmware=%.1f hardware=%.1f\n", info.sw_version / 10.0, info.hw_version / 10.0);
    std::printf("  fclk=%u Hz feature=0x%08X [%s]\n", bus.clockFrequency(), bus.feature(),
                featureString(bus.feature()).c_str());

    std::printf("=== configure (%s) ===\n", args.classic ? "classic CAN" : "CAN FD");
    bus.configure(args.config);
    printTiming("nominal", bus.nominalTiming(), bus.clockFrequency());
    if (bus.isFd()) {
      printTiming("data", bus.dataTiming(), bus.clockFrequency());
    }

    if (args.trigger) {
      std::printf("=== tx trigger ===\n");
      bus.send(makeFrame(args.trigger_id, args.trigger_data, args.config.fd));
      std::printf("  sent %03X [%zu] ", args.trigger_id, args.trigger_data.size());
      for (uint8_t b : args.trigger_data) {
        std::printf("%02X ", b);
      }
      std::printf("\n");
    } else {
      std::printf("=== passive monitor for %.1fs ===\n", args.seconds);
    }

    std::printf("=== rx for %.1fs (echo filtered, showing %s) ===\n", args.seconds,
                args.filter_rx_id ? "0x481" : "all ids");
    const auto deadline =
        std::chrono::steady_clock::now() + std::chrono::duration<double>(args.seconds);
    auto last_trigger = std::chrono::steady_clock::now();
    auto next_report = std::chrono::steady_clock::now() + std::chrono::seconds(2);
    auto window_start = std::chrono::steady_clock::now();

    canfd::CanFrame frame;
    std::map<uint32_t, unsigned> counts;
    std::vector<uint8_t> last_payload;
    bool have_last = false;
    unsigned total = 0;
    unsigned window = 0;
    while (!g_stop.load() && std::chrono::steady_clock::now() < deadline) {
      const auto now = std::chrono::steady_clock::now();
      if (args.trigger && now - last_trigger >= std::chrono::seconds(1)) {
        bus.send(makeFrame(args.trigger_id, args.trigger_data, args.config.fd));
        last_trigger = now;
      }

      if (!bus.receive(frame, std::chrono::milliseconds(100))) {
        continue;
      }
      if (frame.echo) {
        continue;
      }
      if (args.filter_rx_id && frame.id != args.rx_id) {
        continue;
      }
      ++counts[frame.id];
      ++total;
      ++window;

      const bool changed =
          !have_last || frame.size != last_payload.size() ||
          !std::equal(frame.data.begin(), frame.data.begin() + frame.size, last_payload.begin());
      if (args.print_all || changed) {
        std::printf("  %s\n", frame.toString().c_str());
        last_payload.assign(frame.data.begin(), frame.data.begin() + frame.size);
        have_last = true;
      }

      if (now >= next_report) {
        const double elapsed =
            std::chrono::duration<double>(now - window_start).count();
        std::printf("--- %u frames in %.1fs (%.0f fps) ---\n", window, elapsed,
                    elapsed > 0.0 ? window / elapsed : 0.0);
        window = 0;
        window_start = now;
        next_report = now + std::chrono::seconds(2);
      }
    }

    std::printf("=== summary: %u frames ===\n", total);
    for (const auto& entry : counts) {
      std::printf("  id %03X : %u\n", entry.first, entry.second);
    }
  } catch (const std::exception& exc) {
    std::fprintf(stderr, "error: %s\n", exc.what());
    bus.close();
    return 1;
  }

  bus.close();
  return 0;
}

File: python/examples/canfd_demo.py
"""End-to-end hardware demo for gsusb-canfd (Python), mirroring examples/canfd_demo.cpp.

It scans for an adapter, opens it, prints the negotiated bit timing, transmits a
trigger frame on 0x501 and prints every response it sees (0x481 by default),
skipping the adapter's own TX echo.

Install the package first (``pip install gsusb-canfd`` or ``pip install ./python``)
then run:

    python examples/canfd_demo.py --seconds 6
    python examples/canfd_demo.py --monitor --seconds 5
    python examples/canfd_demo.py --classic --id 123 --rx-id 456 --take-data 11223344
"""

from __future__ import annotations

import argparse
import time
from typing import List, Optional

from gsusb_canfd import BusConfig, CanFdBus, CanFrame, scan_adapters
from gsusb_canfd.protocol import (
    FEATURE_BT_CONST_EXT,
    FEATURE_FD,
    FEATURE_HW_TIMESTAMP,
    FEATURE_LISTEN_ONLY,
    FEATURE_LOOPBACK,
    FEATURE_ONE_SHOT,
    CanFdError,
)

FEATURES = (
    (FEATURE_LISTEN_ONLY, "listen_only"),
    (FEATURE_LOOPBACK, "loopback"),
    (FEATURE_ONE_SHOT, "one_shot"),
    (FEATURE_HW_TIMESTAMP, "hw_timestamp"),
    (FEATURE_FD, "fd"),
    (FEATURE_BT_CONST_EXT, "bt_const_ext"),
)


def feature_string(feature: int) -> str:
    names = [name for bit, name in FEATURES if feature & bit]
    return "|".join(names) if names else "none"


def parse_hex_bytes(text: str) -> bytes:
    cleaned = "".join(ch for ch in text if ch.isalnum())
    return bytes.fromhex(cleaned) if cleaned else b""


def make_frame(identifier: int, data: bytes, fd: bool) -> CanFrame:
    return CanFrame(
        id=identifier,
        data=data,
        extended=identifier > 0x7FF,
        fd=fd,
        brs=fd,
    )


def sample_point(timing) -> float:
    total = 1 + timing.tseg1 + timing.phase_seg2
    return (1 + timing.tseg1) / total


def bitrate(timing, fclk: int) -> float:
    total = 1 + timing.tseg1 + timing.phase_seg2
    return fclk / (timing.brp * total)


def print_timing(label: str, timing, fclk: int) -> None:
    print(f"  {label:<8} brp={timing.brp} tseg1={timing.tseg1} "
          f"tseg2={timing.phase_seg2} sjw={timing.sjw}  ->  "
          f"{bitrate(timing, fclk):.0f} bit/s @ {sample_point(timing) * 100:.1f}%")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="gsusb-canfd hardware demo")
    parser.add_argument("--seconds", type=float, default=6.0, help="how long to receive")
    parser.add_argument("--monitor", action="store_true",
                        help="do not trigger, just listen")
    parser.add_argument("--no-trigger", action="store_true", help="do not send the trigger")
    parser.add_argument("--classic", action="store_true", help="use classic CAN only")
    parser.add_argument("--bitrate", type=int, default=1_000_000)
    parser.add_argument("--sample-point", type=float, default=0.80)
    parser.add_argument("--data-bitrate", type=int, default=5_000_000)
    parser.add_argument("--data-sample-point", type=float, default=0.75)
    parser.add_argument("--id", type=lambda v: int(v, 16), default=0x501,
                        help="trigger arbitration id (hex)")
    parser.add_argument("--take-data", type=parse_hex_bytes, default=b"\x00\x02\x50\x01",
                        help="trigger payload (hex)")
    parser.add_argument("--rx-id", type=lambda v: int(v, 16), default=0x481,
                        help="only show this id (hex)")
    parser.add_argument("--all", action="store_true", help="show every id, not just --rx-id")
    parser.add_argument("--all-frames", action="store_true",
                        help="print repeated payloads too")
    return parser


def main(argv: Optional[List[str]] = None) -> int:
    args = build_parser().parse_args(argv)
    trigger = not (args.monitor or args.no_trigger)
    fd = not args.classic

    print("=== scan ===")
    for index, adapter in enumerate(scan_adapters(0, 0)):
        print(f"  {adapter.name()} | {adapter.vendor_id:04X}:{adapter.product_id:04X} "
              f"bus={adapter.bus} addr={adapter.address} serial={adapter.serial}")

    bus = CanFdBus()
    try:
        print("=== open ===")
        bus.open()
        info = bus.device_info
        print(f"  endpoints in=0x{bus.ep_in:02X} out=0x{bus.ep_out:02X}")
        if info is not None:
            print(f"  firmware={info.sw_version / 10.0:.1f} "
                  f"hardware={info.hw_version / 10.0:.1f}")
        print(f"  fclk={bus.fclk_can} Hz feature=0x{bus.feature:08X} "
              f"[{feature_string(bus.feature)}]")

        print(f"=== configure ({'classic CAN' if args.classic else 'CAN FD'}) ===")
        bus.configure(BusConfig(
            bitrate=args.bitrate,
            sample_point=args.sample_point,
            data_bitrate=args.data_bitrate,
            data_sample_point=args.data_sample_point,
            fd=fd,
        ))
        print_timing("nominal", bus.nominal_timing, bus.fclk_can)
        if bus.is_fd and bus.data_timing is not None:
            print_timing("data", bus.data_timing, bus.fclk_can)

        if trigger:
            print("=== tx trigger ===")
            bus.send(make_frame(args.id, args.take_data, fd))
            print(f"  sent {args.id:03X} [{len(args.take_data)}] "
                  f"{' '.join(f'{b:02X}' for b in args.take_data)}")
        else:
            print(f"=== passive monitor for {args.seconds:.1f}s ===")

        target = "all ids" if args.all else f"0x{args.rx_id:03X}"
        print(f"=== rx for {args.seconds:.1f}s (echo filtered, showing {target}) ===")
        deadline = time.monotonic() + args.seconds
        last_trigger = time.monotonic()
        next_report = time.monotonic() + 2.0
        window_start = time.monotonic()

        counts = {}
        last_payload: Optional[bytes] = None
        total = 0
        window = 0
        while time.monotonic() < deadline:
            now = time.monotonic()
            if trigger and now - last_trigger >= 1.0:
                bus.send(make_frame(args.id, args.take_data, fd))
                last_trigger = now

            frame = bus.receive(timeout=0.1)
            if frame is None or frame.echo:
                continue
            if not args.all and frame.id != args.rx_id:
                continue

            counts[frame.id] = counts.get(frame.id, 0) + 1
            total += 1
            window += 1

            if args.all_frames or frame.data != last_payload:
                print(f"  {frame}")
                last_payload = frame.data

            if now >= next_report:
                elapsed = now - window_start
                fps = window / elapsed if elapsed > 0 else 0.0
                print(f"--- {window} frames in {elapsed:.1f}s ({fps:.0f} fps) ---")
                window = 0
                window_start = now
                next_report = now + 2.0

        print(f"=== summary: {total} frames ===")
        for identifier, count in sorted(counts.items()):
            print(f"  id {identifier:03X} : {count}")
    except KeyboardInterrupt:
        print("\ninterrupted")
    except CanFdError as exc:
        print(f"error: {exc}")
        bus.close()
        return 1
    finally:
        bus.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
